test/test_cpp_api_parity.py

```diff
@@ -0,0 +1,456 @@
+    def _test_torch_nn_module_ctor_args(self, module_name):
+        python_module_class = getattr(torch.nn, module_name)
+        module_metadata = torch_nn_modules.module_metadata_map[module_name]
+        cpp_default_constructor_args = module_metadata['cpp_default_constructor_args']
+
+        cpp_module_option = 'torch::nn::{}Options{}'.format(module_name, cpp_default_constructor_args)
+        if PY2:
+            init_arg_spec = inspect.getargspec(python_module_class.__init__)
+        else:
+            init_arg_spec = inspect.getfullargspec(python_module_class.__init__)
+        init_kwargs_defaults = init_arg_spec.defaults
+        init_kwargs = init_arg_spec.args[-len(init_kwargs_defaults):]
+        for arg_name, python_default_value in zip(init_kwargs, init_kwargs_defaults):
+            cpp_module_option += '.{}({})'.format(arg_name, self._python_arg_to_cpp_arg(python_default_value).value)
+
+        cpp_sources = TORCH_NN_MODULE_COMMON_TEST_HARNESS + module_metadata.get('cpp_sources', '')
+        cpp_sources += TORCH_NN_MODULE_TEST_CTOR_ARGS.substitute(
+            module_name=module_name,
+            module_qualified_name='torch::nn::{}'.format(module_name),
+            module_option=cpp_module_option)
+        cpp_test_name = module_name + '_test_ctor_args'
+        cpp_module = self._compile_cpp_code_inline(
+            name=cpp_test_name, cpp_sources=cpp_sources, functions=cpp_test_name)
+
+        getattr(cpp_module, cpp_test_name)()
+
+    def _test_torch_nn_module_variant(self, test_params):
+        torch_nn_test_methods = ['init', 'forward', 'backward']
+
+        def generate_test_cpp_sources(test_params, template, extra_stmts):
+            example_inputs = test_params.example_inputs
+            input_arg_types = [self._python_arg_to_cpp_arg(arg).type for arg in example_inputs]
+            input_arg_declarations = ',\n'.join(
+                ['{} arg{}'.format(input_arg_types[i], str(i)) for i in range(len(input_arg_types))])
+            input_args = ',\n'.join(['arg{}'.format(str(i)) for i in range(len(input_arg_types))])
+            test_cpp_sources = template.substitute(
+                module_variant_name=test_params.module_variant_name,
+                module_qualified_name='torch::nn::{}'.format(test_params.module_name),
+                cpp_constructor_args=test_params.cpp_constructor_args,
+                input_arg_declarations=input_arg_declarations,
+                input_args=input_args,
+                extra_stmts=extra_stmts)
+            return test_cpp_sources
+
+        def setup_init_test(test_params):
+            # We are generating the attribute equality checks manually here,
+            # because it is not possible to have a `.attributes()` API that returns
+            # non-parameter / non-buffer attributes in a C++ torch::nn module.
+            def generate_attr_equality_checks(module, stmts,
```

Comment:
It's kind of weird to be modifying `stmts`. I think it would be nicer (and easier to understand) if this function returned a List instead of modifying that list in-place.